bindings/java/src/org/sleuthkit/datamodel/TaggingManager.java

```diff
@@ -0,0 +1,459 @@
+			connection.beginTransaction();
+
+			String queryTemplate = "UPDATE tag_names SET tag_set_id = %d where tag_name_id = %d";
+			stmt.executeUpdate(String.format(queryTemplate, tagSet.getId(), tagName.getId()));
+
+			connection.commitTransaction();
+		} catch (SQLException ex) {
+			connection.rollbackTransaction();
+			throw new TskCoreException(String.format("Error adding TagName (id=%d) to TagSet (id=%s)", tagName.getId(), tagSet.getId()), ex);
+		} finally {
+			connection.close();
+			skCase.releaseSingleUserCaseWriteLock();
+		}
+	}
+
+	/**
+	 * Inserts a row into the blackboard_artifact_tags table in the case
+	 * database.
+	 *
+	 * @param artifact The blackboard artifact to tag.
+	 * @param tagName  The name to use for the tag.
+	 * @param comment  A comment to store with the tag.
+	 *
+	 * @return A BlackboardArtifactTag data transfer object (DTO) for the new
+	 *         row.
+	 *
+	 * @throws TskCoreException
+	 */
+	public BlackboardArtifactTagChange addArtifactTag(BlackboardArtifact artifact, TagName tagName, String comment) throws TskCoreException {
+		if (artifact == null || tagName == null) {
+			throw new IllegalArgumentException("NULL argument passed to addArtifactTag");
+		}
+
+		CaseDbConnection connection = skCase.getConnection();
+		skCase.acquireSingleUserCaseWriteLock();
+		List<BlackboardArtifactTag> removedTags = new ArrayList<>();
+		try {
+			connection.beginTransaction();
+			// If a TagName is part of a TagSet remove any existing tags from the
+			// set that are currenctly on the artifact
+			int tagSetId = tagName.getTagSetId();
+			if (tagSetId > 0) {
+				// Get the list of all of the blackboardArtifactTags that use
+				// TagName for the given artifact.
+				String selectQuery = String.format("SELECT * from blackboard_artifact_tags JOIN tag_names ON tag_names.tag_name_id = blackboard_artifact_tags.tag_name_id JOIN tsk_examiners on tsk_examiners.examiner_id = blackboard_artifact_tags.examiner_id WHERE artifact_id = %d AND tag_names.tag_set_id = %d", artifact.getId(), tagSetId);
+
+				try (Statement stmt = connection.createStatement(); ResultSet resultSet = stmt.executeQuery(selectQuery)) {
+					while (resultSet.next()) {
+						BlackboardArtifactTag bat
+								= new BlackboardArtifactTag(resultSet.getLong("tag_id"),
+										artifact,
+										skCase.getContentById(artifact.getObjectID()),
+										tagName,
+										resultSet.getString("comment"),
+										resultSet.getString("login_name"));
+
+						removedTags.add(bat);
+					}
+				}
+
+				if (!removedTags.isEmpty()) {
+					// Remove the tags.
+					String removeQuery = String.format("DELETE FROM blackboard_artifact_tags WHERE tag_id IN (SELECT tag_id FROM blackboard_artifact_tags JOIN tag_names ON tag_names.tag_name_id = blackboard_artifact_tags.tag_name_id WHERE artifact_id = %d AND tag_names.tag_set_id = %d)", artifact.getId(), tagSetId);
+					try (Statement stmt = connection.createStatement()) {
```

Comment:
I think we already have all of the tag IDs in hand from the previous query and can just make a CSV-list of the IDs instead of doing another SELECT for the IN clause. Or is this just being careful about changes that might have occurred since the previous query?

Reply:
The original query was a little simpler, but didn't work, this was the solution on found on the internet.  I hadn't though about the fact I already had the ids.  I will change it to a comma separated list of ideas.  Good idea!